### PR TITLE
reflection_pad2d support

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #77854
* #77816
* #77499
* #77740
* #77739
* __->__ #77681

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

### CODE REVIEW

File: torch/_meta_registrations.py [hunk 1]
```diff
@@ -114,3 +114,32 @@ def meta_adaptive_avg_pool2d(self, output_size):
...
+    output_h = input_h + pad_t + pad_b
+    output_w = input_w + pad_l + pad_r
+
+    if self.ndim == 3:
+        return self.new_empty((nplane, output_h, output_w))
+    else:
+        return self.new_empty((nbatch, nplane, output_h, output_w))
+
+
+    return self.new_empty(self.shape[:-3] + tuple(output_size))
```
Comment (ngimel) [inline torch/_meta_registrations.py:143]: you can never get to this line?

File: torch/_meta_registrations.py [hunk 2]
```diff
@@ -114,3 +114,32 @@ def meta_adaptive_avg_pool2d(self, output_size):
 def meta_adaptive_avg_pool3d(self, output_size):
     check(self.ndim == 4 or self.ndim == 5, f"Expected 4D or 5D tensor, but got {self.shape}")
     return self.new_empty(self.shape[:-3] + tuple(output_size))
+
+@torch.library.impl(meta_lib, "reflection_pad2d")
+def meta_reflection_pad2d(self, padding):
```
Comment (ngimel) [inline torch/_meta_registrations.py:119]: this will work for various padding modes, so call it `meta_pad2d` to future-proof?

Comment (ezyang) [reply torch/_meta_registrations.py:119]: oddly enough, none of the other padding modes have xfails in test_meta.py, maybe cuz they all got ported to structured

**Review by mruberry — APPROVED** [review]
Stamped!

**Pull request closed without merging**